File: src/main/java/com/zooplus/jacekb/learningTime/akka/pi/cluster/PiClusterClient.java
package com.zooplus.jacekb.learningTime.akka.pi.cluster;

import akka.actor.*;
import akka.contrib.pattern.ClusterClient;
import akka.event.Logging;
import akka.event.LoggingAdapter;
import akka.japi.pf.ReceiveBuilder;
import com.typesafe.config.Config;
import com.typesafe.config.ConfigFactory;
import com.zooplus.jacekb.learningTime.akka.pi.java.Calculate;
import com.zooplus.jacekb.learningTime.akka.pi.java.PiApproximation;

import java.util.HashSet;
import java.util.Set;

/**
 * @author Jacek Bilski
 * @version $Revision$
 *          $Id$
 */
public class PiClusterClient extends AbstractActor {

    LoggingAdapter log = Logging.getLogger(getContext().system(), this);

	public static void main(String[] args) {
		Config config = ConfigFactory.load("clusterClient");
		ActorSystem system = ActorSystem.create("PiClusterSystem", config);
		system.actorOf(Props.create(PiClusterClient.class), "piClusterClient");
	}

    public PiClusterClient() {
        Set<ActorSelection> initialContacts = new HashSet<>();
        initialContacts.add(getContext().actorSelection("akka.tcp://PiClusterSystem@localhost:2553/user/receptionist"));
        ActorRef clusterClient = getContext().actorOf(ClusterClient.defaultProps(initialContacts));
        clusterClient.tell(new ClusterClient.Send("/user/manager", new Calculate()), self());
        receive(ReceiveBuilder.
                match(PiApproximation.class, this::processResult).
                build());
    }

    private void processResult(PiApproximation piApproximation) {
        log.info("\n\tPi approximation: \t"+piApproximation.getPi()+"\n\tCalculation time: \t"+piApproximation.getDuration());
        getContext().system().shutdown();
    }
}

File: src/main/scala/com/zooplus/jacekb/learningTime/akka/cluster/PiClusterClient.scala
package com.zooplus.jacekb.learningTime.akka.cluster

import akka.actor._
import com.zooplus.jacekb.learningTime.akka.pi.Commons.{Calculate, PiApproximation}
import com.typesafe.config.ConfigFactory
import akka.contrib.pattern.ClusterClient

/**
 * Created with IntelliJ IDEA.
 * User: jacek_bilski
 * Date: 11/07/13
 * Time: 08:35
 */
object PiClusterClient {

	def main(args: Array[String]) {
		val config = ConfigFactory.load("clusterClient")
		val system = ActorSystem("ClusterSystem", config)
		system.actorOf(Props[Client], name = "client")
	}

	class Client extends Actor with ActorLogging {

//		val manager = context.actorSelection("/user/manager")
		val initialContacts = Set(context.actorSelection("akka.tcp://ClusterSystem@localhost:2553/user/receptionist"))

		val c = context.actorOf(ClusterClient.props(initialContacts))
		c ! ClusterClient.Send("/user/manager", Calculate, localAffinity = true)
		println("Sent calculate message")

		def receive = {
			case PiApproximation(pi, duration) ⇒
				log.info("\n\tPi approximation: \t\t%s\n\tCalculation time: \t%s".format(pi, duration))
				context.system.shutdown()
		}
	}

}
